sdk/nodejs/output.ts

```diff
@@ -268,23 +295,36 @@ async function applyHelperAsync<T, U>(value: T, isKnown: boolean, isSecret: bool
         }
     }
 
+    // If we are running with unknown values and the value is explicitly unknown but does not actually
+    // contain any unknown values, collapse its value to the unknown value. This ensures that callbacks
+    // that expect to see unknowns in outputs that are not known will always do so.
+    if (runWithUnknowns && !isKnown && !containsUnknowns(value)) {
+        value = <T><any>unknown;
+    }
+
     const transformed = await func(value);
     if (Output.isInstance(transformed)) {
         // Note: if the func returned a Output, we unwrap that to get the inner value returned by
         // that Output.  Note that we are *not* capturing the Resources of this inner Output.
         // That's intentional.  As the Output returned is only supposed to be related this *this*
         // Output object, those resources should already be in our transitively reachable resource
         // graph.
+        //
+        // Note that the result is only known if the inner result is also known. We will allow the
+        // outer result to be unknown if the caller has explicitly requested that the callback run
+        // even with unknown values. This allows a callback to process an unknown value and return
+        // a known value. This is necessary for the proxy to return known values for properties in
+        // objects that contain unknowns.
 
         // Note: we intentionally await all the promises of the transformed value.  This way we
         // properly propogate any rejections of any of them through ourselves as well.
-        const innerValue = await transformed.promise();
+        const innerValue = <U>(await (<any>transformed).promise(/*withUnknowns*/ true));
```

Comment:
why are we not passing `runWithUnknowns` along here as the arg?  It feels a bit strange to me to unilaterally pass along true here.

Reply:
nit: avoid casting to `<any>` if you can, cast to `OutputInstance` as that will at least allow TS to state "there is a `promise` property here that takes an optional bool".

Reply:
We want to preserve any nested unknowns in the result, so we need to pass `true` for `withUnknowns`.

Reply:
> cast to OutputInstance as that will at least allow TS to state "there is a promise property here that takes an optional bool".

Can do. That cast is purely a compile-time construct, right?

Reply:
correct.  it's just to get reasonable ts behavior.  `<any>` casts are generally a 'last resort' when all else fails issue.